skybar/src/main/java/org/wtf/skybar/registry/SkybarRegistry.java

```diff
@@ -23,106 +23,77 @@
-     */
-    @GuardedBy("this")
-    private final List<DeltaListener> listenersCopy = new ArrayList<>();
-
-    /**
-     * Called each time a line is visited
-     */
-    public void visitLine(long index) {
-        long l = phaser.writerCriticalSectionEnter();
-        try {
-            activeCounts.compute(index, (count, adder) -> {
-                if (adder == null) {
-                    LongAdder longAdder = new LongAdder();
-                    longAdder.increment();
-                    return longAdder;
-                }
-
-                adder.increment();
-
-                return adder;
-            });
-        } finally {
-            phaser.writerCriticalSectionExit(l);
-        }
-    }
+    private final List<DeltaListener> listeners = new CopyOnWriteArrayList<>();
 
     /**
      * Can only be called once for a given source / line pair
      *
      * @param sourceName path to source file
      * @param lineNumber line number
-     * @return index to be used with visitLine
      */
-    public long registerLine(String sourceName, int lineNumber) {
-        long l = phaser.writerCriticalSectionEnter();
-        long index;
-
-        try {
-            index = nextIndex.incrementAndGet();
-            indexToSourceLoc.put(index, new SourceLocation(sourceName, lineNumber));
-        } finally {
-            phaser.writerCriticalSectionExit(l);
+    public void registerLine(String sourceName, int lineNumber) {
+        Map<Integer, LongAdder> lines = visits.get(sourceName);
+        if(lines == null) {
+            Map<Integer, LongAdder> newLines = new HashMap<>();
+            Map<Integer, LongAdder> existingLines = visits.putIfAbsent(sourceName, newLines);
+            lines = existingLines != null ? existingLines : newLines;
         }
-
-        return index;
+        lines.putIfAbsent(lineNumber, new LongAdder());
     }
 
     /**
-     * @param deltaBuffer map to use to write delta into
+     * Can only be called after registerLine has been called for the same line
+     * @param sourceName name of the source file
+     * @param lineNumber the line number to get the LongAdder for
+     * @return the LongAdder used to accumulate visit counts for the line
      */
-    public void updateListeners(Map<String, IntLongMap> deltaBuffer) {
-        deltaBuffer.forEach((s, counts) -> counts.clear());
-
-        synchronized (this) {
-            flipPhase();
-
-            inactiveCounts.forEach((index, adder) -> {
-
-            /*
-             this is accessed outside the registry's phaser read lock, but all keys we will access here for the
-             delta must already have been added at registration time
-            */
-                SourceLocation location = indexToSourceLoc.get(index);
+    public LongAdder getAdderForLine(String sourceName, int lineNumber) {
+        return visits.get(sourceName).get(lineNumber);
```

Comment:
perhaps throw IllegalStateException if the state set by registerLine isn't there to make errors in usage easier to track down? Not that we really need to worry about other people using this...